--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -47,30 +_,6 @@
     }
 
     @Override
-    protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        BlockPos blockPos = pos.above();
-        if (level.isEmptyBlock(blockPos)) {
-            int i = 1;
-
-            while (level.getBlockState(pos.below(i)).is(this)) {
-                i++;
-            }
-
-            if (i < 3) {
-                int ageValue = state.getValue(AGE);
-                if (ageValue == 15) {
-                    level.setBlockAndUpdate(blockPos, this.defaultBlockState());
-                    BlockState blockState = state.setValue(AGE, Integer.valueOf(0));
-                    level.setBlock(pos, blockState, 4);
-                    level.neighborChanged(blockState, blockPos, this, null, false);
-                } else {
-                    level.setBlock(pos, state.setValue(AGE, Integer.valueOf(ageValue + 1)), 4);
-                }
-            }
-        }
-    }
-
-    @Override
     protected VoxelShape getCollisionShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
         return COLLISION_SHAPE;
     }
@@ -113,7 +_,8 @@
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity) {
-        entity.hurt(level.damageSources().cactus(), 1.0F);
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
+        entity.hurt(level.damageSources().cactus().eventBlockDamager(level, pos), 1.0F); // CraftBukkit
     }
 
     @Override
