--- a/net/minecraft/world/level/block/MangrovePropaguleBlock.java
+++ b/net/minecraft/world/level/block/MangrovePropaguleBlock.java
@@ -121,19 +_,6 @@
     }
 
     @Override
-    protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (!isHanging(state)) {
-            if (random.nextInt(7) == 0) {
-                this.advanceTree(level, pos, state, random);
-            }
-        } else {
-            if (!isFullyGrown(state)) {
-                level.setBlock(pos, state.cycle(AGE), 2);
-            }
-        }
-    }
-
-    @Override
     public boolean isValidBonemealTarget(LevelReader level, BlockPos pos, BlockState state) {
         return !isHanging(state) || !isFullyGrown(state);
     }
