From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: WarlordWeaponry <minecraft2013@cox.net>
Date: Sun, 24 Nov 2024 22:51:46 -0500
Subject: [PATCH] sf-Fix-Falling-Particle-Handlers


diff --git a/src/main/java/com/destroystokyo/paper/event/block/BlockFallingParticleEvent.java b/src/main/java/com/destroystokyo/paper/event/block/BlockFallingParticleEvent.java
index 602c40448f2ff02ad45af4d529fcd67d6e379190..3d1bb5a6af0785ec56ba9a4df19f35b2bd1f9a3a 100644
--- a/src/main/java/com/destroystokyo/paper/event/block/BlockFallingParticleEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/block/BlockFallingParticleEvent.java
@@ -9,6 +9,8 @@ import org.jetbrains.annotations.NotNull;
 
 public class BlockFallingParticleEvent extends Event implements Cancellable {
 
+	private static final HandlerList HANDLER_LIST = new HandlerList();
+
 	private final Block mBlock;
 	private final Location mLocation;
 	public BlockFallingParticleEvent(Block block) {
@@ -32,7 +34,12 @@ public class BlockFallingParticleEvent extends Event implements Cancellable {
 
 	}
 
-	@Override public @NotNull HandlerList getHandlers() {
-		return null;
+	@Override
+	public HandlerList getHandlers() {
+		return HANDLER_LIST;
+	}
+
+	public static HandlerList getHandlerList() {
+		return HANDLER_LIST;
 	}
 }
